python/tk_houdini/ui_generation.py

```diff
@@ -436,22 +436,39 @@ def create_panels(self, panels_file):
 
         # get the cmds that launch panels so we can get additional info
         # about the panels when we need it.
-        cmds_by_panel_callback = {}
+        cmds_by_app_inst = {}
         for cmd in self._commands:
             if not cmd.get_type() == "panel":
                 continue
-            cmds_by_panel_callback[cmd.callback] = cmd
+            app_cmds = cmds_by_app_inst.setdefault(cmd.get_app_instance_name(), [])
+            app_cmds.append(cmd)
 
         for panel_cmd in self._panel_commands:
 
-            if not panel_cmd.callback in cmds_by_panel_callback:
-                # currently we rely on a menu command to be registered 
-                # for each panel in order to get the information we need
-                # to display the panel in the UI. If there is no corresponding
-                # command, don't show the panel.
+            # match the panel command with a registered command to get the
+            # info needed to display the panel in the UI.
+
+            if not panel_cmd.get_app_name():
+                # we need to find the matching command and the app is required
+                # for that. if we can't get the app name, just continue
                 continue
 
-            launch_cmd = cmds_by_panel_callback[panel_cmd.callback]
+            app_inst_name = panel_cmd.get_app_instance_name()
+            app_cmds = cmds_by_app_inst[app_inst_name]
+
+            if len(app_cmds) == 1:
+                # assume this is the right command
+                launch_cmd = app_cmds[0]
+            else:
+                # multiple registered panel apps. Until there is a reliable way
+                # to map registered panels to registered commands, not sure
+                # there is a good way to make this work in houdini. just log
+                # the troubles for now.
+                self._engine.log_debug(
+                    "Multiple registered panel commands for app instance '%s'. "
+                    "Unable to build the panel." % (app_inst_name,)
+                )
```

Comment:
What it boils down to is the need to build the panel ahead of time during the `post_app_init` in order to write the xml file. At that time, I need the `unique_id` of the panel which I get from the registered panels, but I also need access to the info available in the corresponding menu command (display name, icon, description, help, etc.). Those things aren't necessarily available in the registered panel info, but most are always in the registered command (some smart defaulting is handled by core itself). Both the `unique_id` and the command info gets embedded into the xml file for displaying the panel and when `show_panel` is called, the interface is retrieved via the `unique_id` and shown (it has to be since that's what core provides via `show_panel`). 

The confusing part, I think, is that there are panel commands registered, but there's no clean way to map those register commands of type `panel` back to the `unique_id` returned form the `register_panel` method. I really don't like the old approach or the new one, and it would be good to have a discussion with @manneohrstrom about this at some point.

One solution might be to have `register_panel` automatically register the command as well, and add the `unique_id` to the properties dict that comes with the command. Right now they're called separately:

https://github.com/shotgunsoftware/tk-multi-shotgunpanel/blob/master/app.py#L36-L47

The signatures of those methods is similar, and it feels odd to me to have to do both. Of course, maybe I'm missing something obvious...